test/jdk/sun/awt/image/gif/GifComparison.java

```diff
@@ -0,0 +1,431 @@
+import java.awt.image.ColorModel;
+import java.awt.image.ImageConsumer;
+import java.awt.image.IndexColorModel;
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.nio.file.*;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.Hashtable;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * This compares the ImageIO rendering of a gif with the ToolkitImage
+ * rendering.
+ * <p>
+ * This is intended to serve as a helper class for more specific test cases.
+ * However this does have its own main method that inspects a folder of gifs.
+ */
+public class GifComparison {
+
+    /**
+     * This inspects a folder and calls {@link #run(URL)} for each gif file.
+     */
+    public static void main(String[] args) throws Exception {
+        AtomicInteger successCtr = new AtomicInteger();
+        AtomicInteger failureCtr = new AtomicInteger();
+        File dir = new File(args[0]);
+        Files.walkFileTree(dir.toPath(), new FileVisitor<>() {
+            @Override
+            public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) {
+                return FileVisitResult.CONTINUE;
+            }
+
+            @Override
+            public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
+                String s = file.toString();
+                if (s.endsWith(".gif")) {
+                    try {
+                        run(file.toFile().toURI().toURL());
+                        successCtr.incrementAndGet();
+                    } catch (Throwable t) {
+                        failureCtr.incrementAndGet();
+                        t.printStackTrace();
+                    }
+                }
+                return FileVisitResult.CONTINUE;
+            }
+
+            @Override
+            public FileVisitResult visitFileFailed(Path file, IOException exc) {
+                return FileVisitResult.CONTINUE;
+            }
+
+            @Override
+            public FileVisitResult postVisitDirectory(Path dir, IOException exc) {
+                return FileVisitResult.CONTINUE;
+            }
+        });
+
+        System.out.println("Done. Successes = " + successCtr.get() +
+                " Failures = " + failureCtr.get());
+    }
+
+    /**
+     * This iterates over every frame a gif and throws an Error / Exception
+     * if ImageIO and ToolkitImage produce different BufferedImage renderings.
+     *
+     * @param srcURL the URL of the image to inspect
+     *
+     * @return the last frame encoded as a TYPE_INT_ARGB image.
+     *         <p>
+     *         Unit tests may further inspect this image to make sure certain
+     *         conditions are met.
+     */
+    public static BufferedImage run(URL srcURL) throws Throwable {
```

Comment:
If we can actually update this helper to check a specific frame is GIF and not run throw all the frame its better.
In this PR we just need the info of 4th frame.

Is this way of checking all frames useful in any other PR's which you have raised?

Reply:
> Is this way of checking all frames useful in any other PR's which you have raised?

No, I don't think so. (I liked the assurance that every frame was identical, but technically that is beyond the scope of each individual unit test.)

I changed this method to only inspect the last frame.

This made the GifComparison's main() class much less useful, so I removed it. So now the GifComparison class is more limited in scope to unit tests -- and it's no longer a generic tool to identify new problems.